apps/condo/domains/organization/utils/accessSchema.js

```diff
@@ -1,239 +1,308 @@
-        to: { id: organizationId },
-        deletedAt: null,
-    })
-    if (!organizationLink) return false
-
-    return checkOrganizationPermissions(userId, organizationLink.from, permissions)
-}
-
-// TODO(nomerdvadcatpyat): use this function where checkRelatedOrganizationPermission and checkOrganizationPermission used together
-async function checkPermissionInUserOrganizationOrRelatedOrganization (userId, organizationId, permission) {
-    return checkPermissionsInUserOrganizationOrRelatedOrganization(userId, organizationId, [permission])
+ * @returns {Promise<void>}
+ */
+async function resetUserEmployeesCache (userId) {
+    const cacheKey = _getUserOrganizationsCacheKey({ id: userId })
+    await _redisClient.del(cacheKey)
 }
 
 /**
- * @param {string} userId
- * @param {string} organizationId
- * @param {string[]} permissions
- * @return {Promise<boolean>}
+ * Information about single organization, in which user is employed, and its child organizations
+ * @typedef {Object} UserOgranizationInfo
+ * @property {Record<string, boolean>} permissions - user permissions in organization
+ * @property {Array<string>} childOrganizations - list of child organizations to which user share permissions via OrganizationLink
  */
-async function checkPermissionsInUserOrganizationOrRelatedOrganization (userId, organizationId, permissions) {
-    if (!userId || !organizationId) return false
 
-    const hasPermissionsInUserOrganization = await checkOrganizationPermissions(userId, organizationId, permissions)
-    if (hasPermissionsInUserOrganization) return true
-    return await checkRelatedOrganizationPermissions(userId, organizationId, permissions)
-}
+/**
+ * Information about all organizations user has access to
+ * @typedef {Object} UserOgranizationsCache
+ * @property {number} dv - cache entry data version
+ * @property {Record<string, UserOgranizationInfo>} organizations - record where keys are organization ids and values are info it
+ */
 
 /**
- * Check that the user has access in each organization
- *
- * @param userId {string} User.id field
- * @param organizationIds {Array<string>} array of objects related organizations
- * @param permission {string} OrganizationEmployeeRole permission key to check for
- * @return {Promise<boolean>}
+ * Obtains user organizations info via caching or sub-querying
+ * @param {{ id: string }} user - user object
+ * @returns {Promise<UserOgranizationsCache>}
+ * @private
  */
-async function checkOrganizationsPermission (userId, organizationIds, permission) {
-    if (!userId || !isArray(organizationIds) || isEmpty(organizationIds)) return false
+async function _getUserOrganizations (user) {
+    const cacheKey = _getUserOrganizationsCacheKey(user)
+
+    if (!DISABLE_USER_ORGANIZATION_CACHING) {
+        const cachedDataString = await _redisClient.get(cacheKey)
 
-    const uniqOrganizationIds = uniq(organizationIds)
+        if (cachedDataString !== null) {
+            return JSON.parse(cachedDataString)
```

Comment:
It’s also good to check the structure

Reply:
I don't think this is necessary at this point. Shape matching usually is not the fastest part of processing. And it's not user-input, so additional validation is redundant. The only way you can mess it up is by redefining it manually in redis, but if you do so - you're probably doing something wrong